File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/metrics/NavigationEvent.java
package com.mapbox.services.android.navigation.v5.navigation.metrics;

import android.os.Build;
import android.os.Parcel;
import android.support.annotation.NonNull;

import com.mapbox.android.telemetry.Event;
import com.mapbox.services.android.navigation.BuildConfig;

/**
 * Base Event class for navigation events, contains common properties.
 */
abstract class NavigationEvent extends Event {
  private static final String OPERATING_SYSTEM = "Android - " + Build.VERSION.RELEASE;
  private final String operatingSystem = OPERATING_SYSTEM;
  private final String device = Build.MODEL;
  private final String sdkVersion = BuildConfig.MAPBOX_NAVIGATION_VERSION_NAME;
  private final String event;
  private final String created;
  private final String applicationState;
  private final String connectivity;
  private final boolean batteryPluggedIn;
  private final int volumeLevel;
  private final int screenBrightness;
  private final int batteryLevel;
  private String startTimestamp;
  private String sdkIdentifier;
  private String sessionIdentifier;
  private String geometry;
  private String profile;
  private String originalRequestIdentifier;
  private String requestIdentifier;
  private String originalGeometry;
  private String audioType;
  private String locationEngine;
  private String tripIdentifier;
  private double lat;
  private double lng;
  private boolean simulation;
  private int absoluteDistanceToDestination;
  private int percentTimeInPortrait;
  private int percentTimeInForeground;
  private int distanceCompleted;
  private int distanceRemaining;
  private int durationRemaining;
  private int eventVersion;
  private int estimatedDistance;
  private int estimatedDuration;
  private int rerouteCount;
  private int originalEstimatedDistance;
  private int originalEstimatedDuration;
  private int stepCount;
  private int originalStepCount;
  private int legIndex;
  private int legCount;
  private int stepIndex;
  private int voiceIndex;
  private int bannerIndex;
  private int totalStepCount;

  NavigationEvent(@NonNull PhoneState phoneState) {
    this.created = phoneState.getCreated();
    this.volumeLevel = phoneState.getVolumeLevel();
    this.batteryLevel = phoneState.getBatteryLevel();
    this.screenBrightness = phoneState.getScreenBrightness();
    this.batteryPluggedIn = phoneState.isBatteryPluggedIn();
    this.connectivity = phoneState.getConnectivity();
    this.audioType = phoneState.getAudioType();
    this.applicationState = phoneState.getApplicationState();
    this.event = getEventName();
  }

  abstract String getEventName();

  String getEvent() {
    return event;
  }

  String getOperatingSystem() {
    return operatingSystem;
  }

  String getDevice() {
    return device;
  }

  String getCreated() {
    return created;
  }

  String getApplicationState() {
    return applicationState;
  }

  String getConnectivity() {
    return connectivity;
  }

  boolean isBatteryPluggedIn() {
    return batteryPluggedIn;
  }

  int getVolumeLevel() {
    return volumeLevel;
  }

  int getScreenBrightness() {
    return screenBrightness;
  }

  int getBatteryLevel() {
    return batteryLevel;
  }

  String getStartTimestamp() {
    return startTimestamp;
  }

  void setStartTimestamp(String startTimestamp) {
    this.startTimestamp = startTimestamp;
  }

  String getSdkIdentifier() {
    return sdkIdentifier;
  }

  void setSdkIdentifier(String sdkIdentifier) {
    this.sdkIdentifier = sdkIdentifier;
  }

  String getSdkVersion() {
    return sdkVersion;
  }

  String getSessionIdentifier() {
    return sessionIdentifier;
  }

  void setSessionIdentifier(String sessionIdentifier) {
    this.sessionIdentifier = sessionIdentifier;
  }

  String getGeometry() {
    return geometry;
  }

  void setGeometry(String geometry) {
    this.geometry = geometry;
  }

  String getProfile() {
    return profile;
  }

  void setProfile(String profile) {
    this.profile = profile;
  }

  String getOriginalRequestIdentifier() {
    return originalRequestIdentifier;
  }

  void setOriginalRequestIdentifier(String originalRequestIdentifier) {
    this.originalRequestIdentifier = originalRequestIdentifier;
  }

  String getRequestIdentifier() {
    return requestIdentifier;
  }

  void setRequestIdentifier(String requestIdentifier) {
    this.requestIdentifier = requestIdentifier;
  }

  String getOriginalGeometry() {
    return originalGeometry;
  }

  void setOriginalGeometry(String originalGeometry) {
    this.originalGeometry = originalGeometry;
  }

  String getAudioType() {
    return audioType;
  }

  void setAudioType(String audioType) {
    this.audioType = audioType;
  }

  String getLocationEngine() {
    return locationEngine;
  }

  void setLocationEngine(String locationEngine) {
    this.locationEngine = locationEngine;
  }

  String getTripIdentifier() {
    return tripIdentifier;
  }

  void setTripIdentifier(String tripIdentifier) {
    this.tripIdentifier = tripIdentifier;
  }

  double getLat() {
    return lat;
  }

  void setLat(double lat) {
    this.lat = lat;
  }

  double getLng() {
    return lng;
  }

  void setLng(double lng) {
    this.lng = lng;
  }

  boolean isSimulation() {
    return simulation;
  }

  void setSimulation(boolean simulation) {
    this.simulation = simulation;
  }

  int getAbsoluteDistanceToDestination() {
    return absoluteDistanceToDestination;
  }

  void setAbsoluteDistanceToDestination(int absoluteDistanceToDestination) {
    this.absoluteDistanceToDestination = absoluteDistanceToDestination;
  }

  int getPercentTimeInPortrait() {
    return percentTimeInPortrait;
  }

  void setPercentTimeInPortrait(int percentTimeInPortrait) {
    this.percentTimeInPortrait = percentTimeInPortrait;
  }

  int getPercentTimeInForeground() {
    return percentTimeInForeground;
  }

  void setPercentTimeInForeground(int percentTimeInForeground) {
    this.percentTimeInForeground = percentTimeInForeground;
  }

  int getDistanceCompleted() {
    return distanceCompleted;
  }

  void setDistanceCompleted(int distanceCompleted) {
    this.distanceCompleted = distanceCompleted;
  }

  int getDistanceRemaining() {
    return distanceRemaining;
  }

  void setDistanceRemaining(int distanceRemaining) {
    this.distanceRemaining = distanceRemaining;
  }

  int getDurationRemaining() {
    return durationRemaining;
  }

  void setDurationRemaining(int durationRemaining) {
    this.durationRemaining = durationRemaining;
  }

  int getEventVersion() {
    return eventVersion;
  }

  void setEventVersion(int eventVersion) {
    this.eventVersion = eventVersion;
  }

  int getEstimatedDistance() {
    return estimatedDistance;
  }

  void setEstimatedDistance(int estimatedDistance) {
    this.estimatedDistance = estimatedDistance;
  }

  int getEstimatedDuration() {
    return estimatedDuration;
  }

  void setEstimatedDuration(int estimatedDuration) {
    this.estimatedDuration = estimatedDuration;
  }

  int getRerouteCount() {
    return rerouteCount;
  }

  void setRerouteCount(int rerouteCount) {
    this.rerouteCount = rerouteCount;
  }

  int getOriginalEstimatedDistance() {
    return originalEstimatedDistance;
  }

  void setOriginalEstimatedDistance(int originalEstimatedDistance) {
    this.originalEstimatedDistance = originalEstimatedDistance;
  }

  int getOriginalEstimatedDuration() {
    return originalEstimatedDuration;
  }

  void setOriginalEstimatedDuration(int originalEstimatedDuration) {
    this.originalEstimatedDuration = originalEstimatedDuration;
  }

  int getStepCount() {
    return stepCount;
  }

  void setStepCount(int stepCount) {
    this.stepCount = stepCount;
  }

  int getOriginalStepCount() {
    return originalStepCount;
  }

  void setOriginalStepCount(int originalStepCount) {
    this.originalStepCount = originalStepCount;
  }

  int getLegIndex() {
    return legIndex;
  }

  void setLegIndex(int legIndex) {
    this.legIndex = legIndex;
  }

  int getLegCount() {
    return legCount;
  }

  void setLegCount(int legCount) {
    this.legCount = legCount;
  }

  int getStepIndex() {
    return stepIndex;
  }

  void setStepIndex(int stepIndex) {
    this.stepIndex = stepIndex;
  }

  int getVoiceIndex() {
    return voiceIndex;
  }

  void setVoiceIndex(int voiceIndex) {
    this.voiceIndex = voiceIndex;
  }

  int getBannerIndex() {
    return bannerIndex;
  }

  void setBannerIndex(int bannerIndex) {
    this.bannerIndex = bannerIndex;
  }

  int getTotalStepCount() {
    return totalStepCount;
  }

  void setTotalStepCount(int totalStepCount) {
    this.totalStepCount = totalStepCount;
  }

  @Override
  public int describeContents() {
    return 0;
  }

  @Override
  public void writeToParcel(Parcel dest, int flags) {
  }
}

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/events/NavigationEvent.kt
package com.mapbox.navigation.core.telemetry.events

import android.os.Build
import androidx.annotation.CallSuper
import com.google.gson.Gson
import com.mapbox.bindgen.Value
import com.mapbox.common.TelemetrySystemUtils
import com.mapbox.navigation.base.internal.metric.MetricEventInternal
import com.mapbox.navigation.core.BuildConfig

/**
 * Base Event class for navigation events, contains common properties.
 *
 * @property driverMode one of [FeedbackEvent.DriverMode]
 * @property sessionIdentifier id of [driverMode]
 * @property startTimestamp start time of [driverMode]
 * @property navigatorSessionIdentifier group id of modes under one Telemetry session
 */
internal abstract class NavigationEvent(
    phoneState: PhoneState,
) : MetricEventInternal {

    private companion object {
        private val OPERATING_SYSTEM = "Android - ${Build.VERSION.RELEASE}"
    }

    /*
     * Don't remove any fields, cause they should match with
     * the schema downloaded from S3. Look at {@link SchemaTest}
     */
    val version = "2.2"
    val operatingSystem: String = OPERATING_SYSTEM
    val device: String? = Build.MODEL
    val sdkVersion: String = BuildConfig.MAPBOX_NAVIGATION_VERSION_NAME
    val created: String = TelemetrySystemUtils.obtainCurrentDate() // Schema pattern
    val volumeLevel: Int = phoneState.volumeLevel
    val batteryLevel: Int = phoneState.batteryLevel
    val screenBrightness: Int = phoneState.screenBrightness
    val batteryPluggedIn: Boolean = phoneState.isBatteryPluggedIn
    val connectivity: String = phoneState.connectivity
    val audioType: String = phoneState.audioType
    val applicationState: String = phoneState.applicationState // Schema minLength 1
    val event: String = getEventName()

    var sdkIdentifier: String? = null

    var navigatorSessionIdentifier: String? = null
    var startTimestamp: String? = null
    var driverMode: String? = null
    var sessionIdentifier: String? = null
    var geometry: String? = null
    var profile: String? = null
    var originalRequestIdentifier: String? = null
    var requestIdentifier: String? = null
    var originalGeometry: String? = null
    var locationEngine: String? = null
    var tripIdentifier: String? = null
    var lat: Double = 0.toDouble()
    var lng: Double = 0.toDouble()
    var simulation: Boolean = false
    var absoluteDistanceToDestination: Int = 0
    var percentTimeInPortrait: Int = 0
    var percentTimeInForeground: Int = 0
    var distanceCompleted: Int = 0
    var distanceRemaining: Int = 0
    var durationRemaining: Int = 0
    var eventVersion: Int = 0
    var estimatedDistance: Int = 0
    var estimatedDuration: Int = 0
    var rerouteCount: Int = 0
    var originalEstimatedDistance: Int = 0
    var originalEstimatedDuration: Int = 0
    var stepCount: Int = 0
    var originalStepCount: Int = 0
    var legIndex: Int = 0
    var legCount: Int = 0
    var stepIndex: Int = 0
    var voiceIndex: Int = 0
    var bannerIndex: Int = 0
    var totalStepCount: Int = 0
    var appMetadata: AppMetadata? = null

    internal abstract fun getEventName(): String

    override fun toJson(gson: Gson): String = gson.toJson(this)

    override val metricName: String
        get() = getEventName()

    @CallSuper
    override fun toValue(): Value {
        val fields = hashMapOf<String, Value>()

        fields["version"] = version.toValue()
        fields["operatingSystem"] = operatingSystem.toValue()
        device?.let { fields["device"] = it.toValue() }
        fields["sdkVersion"] = sdkVersion.toValue()
        fields["created"] = created.toValue()
        fields["volumeLevel"] = volumeLevel.toValue()
        fields["batteryLevel"] = batteryLevel.toValue()
        fields["screenBrightness"] = screenBrightness.toValue()
        fields["batteryPluggedIn"] = batteryPluggedIn.toValue()
        fields["connectivity"] = connectivity.toValue()
        fields["audioType"] = audioType.toValue()
        fields["applicationState"] = applicationState.toValue()
        fields["event"] = event.toValue()

        sdkIdentifier?.let { fields["sdkIdentifier"] = it.toValue() }

        navigatorSessionIdentifier?.let { fields["navigatorSessionIdentifier"] = it.toValue() }
        startTimestamp?.let { fields["startTimestamp"] = it.toValue() }
        driverMode?.let { fields["driverMode"] = it.toValue() }
        sessionIdentifier?.let { fields["sessionIdentifier"] = it.toValue() }
        geometry?.let { fields["geometry"] = it.toValue() }
        profile?.let { fields["profile"] = it.toValue() }
        originalRequestIdentifier?.let { fields["originalRequestIdentifier"] = it.toValue() }
        requestIdentifier?.let { fields["requestIdentifier"] = it.toValue() }
        originalGeometry?.let { fields["originalGeometry"] = it.toValue() }
        locationEngine?.let { fields["locationEngine"] = it.toValue() }
        tripIdentifier?.let { fields["tripIdentifier"] = it.toValue() }
        fields["lat"] = lat.toValue()
        fields["lng"] = lng.toValue()
        fields["simulation"] = simulation.toValue()
        fields["absoluteDistanceToDestination"] = absoluteDistanceToDestination.toValue()
        fields["percentTimeInPortrait"] = percentTimeInPortrait.toValue()
        fields["percentTimeInForeground"] = percentTimeInForeground.toValue()
        fields["distanceCompleted"] = distanceCompleted.toValue()
        fields["distanceRemaining"] = distanceRemaining.toValue()
        fields["durationRemaining"] = durationRemaining.toValue()
        fields["eventVersion"] = eventVersion.toValue()
        fields["estimatedDistance"] = estimatedDistance.toValue()
        fields["estimatedDuration"] = estimatedDuration.toValue()
        fields["rerouteCount"] = rerouteCount.toValue()
        fields["originalEstimatedDistance"] = originalEstimatedDistance.toValue()
        fields["originalEstimatedDuration"] = originalEstimatedDuration.toValue()
        fields["stepCount"] = stepCount.toValue()
        fields["originalStepCount"] = originalStepCount.toValue()
        fields["legIndex"] = legIndex.toValue()
        fields["legCount"] = legCount.toValue()
        fields["stepIndex"] = stepIndex.toValue()
        fields["voiceIndex"] = voiceIndex.toValue()
        fields["bannerIndex"] = bannerIndex.toValue()
        fields["totalStepCount"] = totalStepCount.toValue()
        appMetadata?.let { fields["appMetadata"] = it.toValue() }

        fields.putAll(customFields() ?: emptyMap())

        return Value.valueOf(fields)
    }

    /**
     * Provide custom fields here or `null` if they are not needed
     */
    protected abstract fun customFields(): Map<String, Value>?
}
